src/libraries/Common/tests/System/IO/ConnectedStreams.cs

```diff
@@ -0,0 +1,163 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO
+{
+    /// <summary>Provides support for in-memory producer/consumer streams.</summary>
+    internal sealed class ConnectedStreams
+    {
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional() =>
+            CreateBidirectional(StreamBuffer.DefaultInitialBufferSize);
+
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <param name="initialBufferSize">The initial buffer size to use when storing data in the connection.</param>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional(int initialBufferSize) =>
+            CreateBidirectional(initialBufferSize, int.MaxValue);
```

Comment:
I'm not sure that a default of maxBufferSize == int.MaxValue makes sense. Why is this better than picking a pragmatic default like 16K or similar?

Reply:
To me it's similar to why there's no bound by default on something like a ConcurrentQueue or a BlockingQueue or a Channel, etc., or even a List.  I don't feel strongly about it, but choosing an upper bound like that feels arbitrary to me.

Reply:
The buffer limit seems like mostly a perf tuning thing to me. Your stream is still fully functional regardless of what value you set. It's just that you trade off blocking the producer for memory usage. As such, I'd prefer to pick a pragmatic default.

Channel has different APIs for creating bounded and unbounded, so you have to make an explicit decision there.